File: python/Stack/evaluate_reverse_polish_notation.py
from typing import List

class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        """
        Evaluate the value of an arithmetic expression in Reverse Polish Notation (RPN).
        Approach:
        - Use a stack to process tokens.
        - For each token:
          - If it's a number, push to stack.
          - If it's an operator, pop two numbers, apply the operator, and push the result.
        Time Complexity: O(n), where n is the number of tokens
        Space Complexity: O(n)
        """
        stack = []
        for c in tokens:
            if c == "+":
                stack.append(stack.pop() + stack.pop())
            elif c == "-":
                op2, op1 = stack.pop(), stack.pop()
                stack.append(op1 - op2)
            elif c == "*":
                stack.append(stack.pop() * stack.pop())
            elif c == "/":
                op2, op1 = stack.pop(), stack.pop()
                stack.append(int(op1 / op2))  # Truncate towards zero
            else:
                stack.append(int(c))
        return stack[0]

# -------------------
# Test Cases
# -------------------
def test_eval_rpn():
    s = Solution()
    # Test case 1: Basic
    assert s.evalRPN(["2","1","+","3","*"]) == 9
    # Test case 2: Division truncates towards zero
    assert s.evalRPN(["4","13","5","/","+"]) == 6
    # Test case 3: Negative numbers
    assert s.evalRPN(["10","6","9","3","+","-11","*","/","*","17","+","5","+"]) == 22
    # Test case 4: Single number
    assert s.evalRPN(["42"]) == 42
    # Test case 5: Subtraction
    assert s.evalRPN(["5","3","-"]) == 2
    # Test case 6: Multiplication
    assert s.evalRPN(["3","-4","*"]) == -12
    # Test case 7: Division negative
    assert s.evalRPN(["7","-3","/"]) == -2
    print("All test cases passed!")

if __name__ == "__main__":
    test_eval_rpn()

File: C++/Stack/evaluate_reverse_polish_notation.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    /*
     * Problem: Evaluate Reverse Polish Notation (RPN)
     * 
     * Approach:
     * - Use stack to track numbers during evaluation
     * - For each token:
     *   - If number: push to stack
     *   - If operator: pop two operands, apply operator, push result
     * 
     * Time: O(n) where n is number of tokens
     * Space: O(n) for stack in worst case
     */
    int evalRPN(vector<string>& tokens) {
        stack<int> st;
        
        for (const string& token : tokens) {
            // If token is operator
            if (token == "+" || token == "-" || 
                token == "*" || token == "/") {
                
                // Pop operands in correct order
                int op2 = st.top(); st.pop();
                int op1 = st.top(); st.pop();
                
                // Apply operator and push result
                if (token == "+") {
                    st.push(op1 + op2);
                }
                else if (token == "-") {
                    st.push(op1 - op2);
                }
                else if (token == "*") {
                    st.push(op1 * op2);
                }
                else if (token == "/") {
                    st.push(op1 / op2);  // Integer division
                }
            }
            // If token is number
            else {
                st.push(stoi(token));
            }
        }
        
        return st.top();
    }
};

/*
 * Detailed C++ Syntax Explanation for Beginners:
 * 
 * 1. Stack Operations:
 *    stack<int> st        // Declare integer stack
 *    st.push(val)        // Add value to top
 *    st.top()            // Get top value
 *    st.pop()            // Remove top value
 * 
 * 2. String Operations:
 *    token == "+"        // String comparison
 *    stoi(token)         // Convert string to integer
 * 
 * 3. Range-based For Loop:
 *    for (const string& token : tokens)
 *    - const: token won't be modified
 *    - string&: reference to avoid copying
 *    - tokens: container being iterated
 * 
 * 4. Vector Parameter:
 *    vector<string>& tokens
 *    - Reference to vector of strings
 * 
 * Note: This implementation:
 * - Handles both operators and numbers
 * - Maintains operator precedence through stack
 * - Uses integer division (truncates towards zero)
 */
